### ISSUE TITLE
`use Trait;` misses trait use within submodule

### ISSUE THREAD

**Issue by cramertj** [rust-lang/rust#53682]
I've seen this a bunch of times and always just thought that trait used-ness detection was broken, but it's actually something much more specific:

([playground](https://play.rust-lang.org/?gist=84dce2c7a3d58c2c5f1dc0bfb90726f4&version=stable&mode=debug&edition=2015))

```rust
trait MyTrait {
    fn trait_method(&self) {}
}
impl<T> MyTrait for T {}

pub struct MyStruct;

pub mod outer_mod {
    use super::{MyStruct, MyTrait};
    pub mod inner_mod {
        use super::*;
        pub fn call_trait_method() {
            5i32.trait_method()
        }
        pub fn return_struct() -> MyStruct {
            MyStruct
        }
    }
}

fn main() {
    let _ = outer_mod::inner_mod::call_trait_method();
    let _ = outer_mod::inner_mod::return_struct();
}
```

The code above produces the following warning:
```
warning: unused import: `MyTrait`
 --> src/main.rs:9:27
  |
9 |     use super::{MyStruct, MyTrait};
  |                           ^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default
```

The use of `use`d items through `use super::*;` in the submodule is handled inconsistently for types vs. traits. The use of the struct from the submodule `inner_mod` is detected and counted as a use of the `use` statement in `outer_mod`, but the same is not true of the use of the trait-- rustc misses that it is used via `use super::*;` in `inner_mod` and reports it as unused.

**Comment by ExpHP** [comment]
Dupe of https://github.com/rust-lang/rust/issues/45268 ?

**Comment by RalfJung** [comment]
I just ran into the same problem, my minimized example is
```rust
mod helpers {
    pub trait IntExt {
        fn id(self) -> Self;
    }
    
    impl IntExt for i32 {
        fn id(self) -> Self { self }
    }
}

use helpers::IntExt;

pub mod user {
    use super::*;
    
    pub fn foo(x: i32) -> i32 {
        x.id()
    }
}
```

**Comment by jespersm** [comment]
Yes, this is the same as #45268, which is fixed on master now.
